Tasks that use jdeps analyzer

esword opened the pull request:
## Before this PR
Maven dependency analyzer has several shortcomings, including those that lead to:
https://github.com/palantir/gradle-baseline/issues/847
https://github.com/palantir/gradle-baseline/issues/614

Plus, the current analyzer tasks cache everything which can lead to OoMs with large projects.

## After this PR
==COMMIT_MSG==
New dependency analyzer yields cleaner results and tasks store output in files so better
caching and analysis can be reused for other purposes
==COMMIT_MSG==

## Possible downsides?
Possibly slightly slower since don't cache contents of all jars across projects.

esword committed:
Tasks that use jdeps analyzer

esword commented:
I have not moved over the existing tasks to use the new analysis file yet.  The file is useful on it's own if we want to make that a separate PR, or can do as part of this one.

esword committed:
add in api deps reporting

esword committed:
style fixes

esword committed:
more style fixes

esword committed:
convert implicit dep task to use dep report

esword committed:
more style fixes

esword committed:
temp hack to verify this is what's causing circle to hang

esword committed:
Merge branch 'develop' into feature/jdeps-analyzer

esword committed:
move unused dependency check to reading from dep report

esword committed:
undo temp hack that was checking timeouts on circle

esword committed:
style fixes

esword committed:
adding logging to see why timing out on circle

esword committed:
fix test name

esword committed:
plugin validation fixes

esword committed:
fix back compat with gradle 5.4

esword committed:
temp ignore test until I come back and handle source-only deps

esword committed:
revert changes to existing dependency plugin classes

esword committed:
use classpath with jdeps to get more info

unfortunately, it doesn't list full path to dependency files, so is
useless for determining other projects that deps came from.  Committing
purely so have record of how to do this should I come back to it.

esword committed:
add in source-only filtering

esword committed:
plugin validation fixes

esword committed:
tweak plugin id to be more appropriate

esword committed:
get newer implicit and unused tasks working

also do some cleanup for task generation

esword committed:
fix tests

esword committed:
javadoc cleanup

esword committed:
expose util class for getting task names for other plugins

esword committed:
handle when source directory isn't present

esword committed:
fix bug where wrong dot file was selected and improve name cleanup

esword committed:
works with exact set to of configs analyze

esword committed:
pass around dependency IDs rather than resolved artifacts where possible

The gradle APIs that use resolved items will be deprecated eventually.
so best not to use them.  This also reduces memory overhead.

esword committed:
handle the immutables annotation situation automatically

esword committed:
Merge branch 'develop' into feature/jdeps-analyzer

esword committed:
add change log

esword committed:
quick mod to handle projects with more than one classes dir

esword committed:
correct annotation for report dir.

Gradle task validation would fail if the directory already existed with
the other annotation.

esword committed:
deconflict task name with the one in core gradle

esword committed:
use full classpath configuration to work around errors in gradle 5.x

the classpath configuration has "runtime" varients for artifacts that
allow it to look at classes directories rather than jars.  This avoids
errors because project-dependencies may not always have jars generated.

CRogers commented on gradle-baseline-java/build.gradle:
```diff
@@ -58,6 +61,10 @@ gradlePlugin {
         }
+        baselineDependencyPluginv2 {
+            id = 'com.palantir.baseline-dependencies-v2'
+            implementationClass = 'com.palantir.baseline.plugins.BaselineDependencyPluginv2'
```
We should just replace it entirely if we have confidence it's good. If we have lower confidence we should take a gradle property to excavate it out rather than an entirely new task (this allows us to track adoption and easily excavate the changes).

CRogers commented on gradle-baseline-java/src/main/groovy/com/palantir/baseline/tasks/dependencies/DependencyAnalysisTask.java:
```diff
@@ -0,0 +1,167 @@
+        return report;
+    }
+
+    public static final class ReportContent {
```
I would use immutables here as it will build this class for you and handle the yaml dumping by generating the correct jackson classes as well, meaning we don't need snakeyaml.

CRogers commented on gradle-baseline-java/src/main/groovy/com/palantir/baseline/tasks/dependencies/DependencyAnalyzer.java:
```diff
@@ -0,0 +1,230 @@
+    }
+
+    /**
+     * Dependencies that are declared in one of the given configurations but not used in byte code.  This will only
```
Does it detect things like `Class.forName(...)` or other dynamic class loading?

CRogers commented on gradle-baseline-java/src/main/groovy/com/palantir/baseline/tasks/dependencies/DependencyUtils.java:
```diff
@@ -0,0 +1,164 @@
+import org.slf4j.LoggerFactory;
+import org.yaml.snakeyaml.Yaml;
+
+public final class DependencyUtils {
```
I think most of these public classes should be package-private.

CRogers commented on gradle-baseline-java/src/main/groovy/com/palantir/baseline/tasks/dependencies/DependencyUtils.java:
```diff
@@ -0,0 +1,164 @@
+    }
+
+    static void writeDepReport(File file, DependencyAnalysisTask.ReportContent content) {
+        Yaml yaml = new Yaml();
```
We should try to use jackson rather than snakeyaml here to be consistent with everything else that uses jackson.

esword committed:
better handling of no sources.

Now generates a stub output file so that downstream tasks don't bomb.

esword committed:
fix pathsensitivity for inputs

Looks like ABSOLUTE is relative to the root project

esword committed:
avoid spurious exceptions

esword committed:
properly detect dependency inputs so can detect when task is dirty.

will merge in Analyzer to the task to avoid duplication

esword committed:
combine analyzer into task

esword committed:
turning off extra test for circle

CRogers commented on gradle-baseline-java/build.gradle:
```diff
@@ -17,6 +17,9 @@ dependencies {
     implementation 'com.palantir.javaformat:palantir-java-format-spi'
     implementation 'com.palantir.javaformat:gradle-palantir-java-format'
+    implementation 'com.paypal.digraph:digraph-parser'
+    implementation 'org.zeroturnaround:zt-exec'
```
We should avoid taking dependencies on extra stuff if it's not necessary. The java process API may be nasty but it's workable. Since we put everything into gradle `buildscript` internally there's great capacity to have breaking changes between gradle plugins wreak, especially from something as widely used as baseline.

CRogers commented on gradle-baseline-java/build.gradle:
```diff
@@ -17,6 +17,9 @@ dependencies {
     implementation 'com.palantir.javaformat:gradle-palantir-java-format'
+    implementation 'com.paypal.digraph:digraph-parser'
+    implementation 'org.zeroturnaround:zt-exec'
+    implementation 'org.yaml:snakeyaml'
```
Can we avoid snakeyaml and use jackson everywhere?

CRogers commented on gradle-baseline-java/src/main/groovy/com/palantir/baseline/plugins/BaselineDependencyPluginv2.java:
```diff
@@ -0,0 +1,143 @@
+                    .getSourceSets();
+
+            createTasksForSourceSet(project, sourceSets.getByName(SourceSet.MAIN_SOURCE_SET_NAME));
+            createTasksForSourceSet(project, sourceSets.getByName(SourceSet.TEST_SOURCE_SET_NAME));
```
What about people using custom `integrationTest` or `integTest` source sets?

CRogers replied:
I guess we can perhaps just ignore them.

CRogers commented on gradle-baseline-java/src/main/groovy/com/palantir/baseline/plugins/BaselineDependencyPluginv2.java:
```diff
@@ -0,0 +1,143 @@
+            t.getIgnored().add("javax.annotation:javax.annotation-api");
+            //Projects may apply this in compileOnly to work around
+            //https://github.com/immutables/immutables/issues/291
+            t.getIgnored().add("org.immutables:value::annotations");
```
It might make sense to ignore all dependencies which are in `compileOnly` and also `annotationProcessor` as this happens to more libraries than just immutables, iirc.

CRogers reviewed (commented):
I didn't manage to allocated as much time as I wanted to review this, still need to test it out locally.

esword committed:
fix bug with classes in same package

correct for bug where jdeps doesn't report dependency if referenced
class in same package as current class, even if it's in a different
module.

iamdanfox commented:
I think we may want to resuscitate this because the maven-dependency-analyzer can't parse bytecode produced by java13.
